scripts/pf2e-extempore-effects.js

```diff
@@ -245,7 +305,74 @@ const onUpdateWorldTime_Wrapper = (wrapped, ...args) => {
     }
   }
 
-  wrapped(...args)
+  // If there was an expired effect notification, and the user requested to advance time beyond the expiry time,
+  // let's prompt the user if they want to continue to requested time or stop at the expiry time.
+  if (expireEffectMessageShown && newWorldTime < requestedNewWorldTime) {
+    const dialogTitle = `${localize('.dialog.expired-effects.title')} ${lastExpiryTime}`;
+    let dialogContent = `<span><h6>${localize('.dialog.expired-effects.content.header')}</h6><ul>`;
+
+    for (const effect of effectsExpired) dialogContent += `<li>${effect.name} (${effect.actor.name})</li>`;
+
+    dialogContent += `</ul>${localize('.dialog.expired-effects.content.message-time')} ${formatTime(newWorldTime - oldWorldTime)}.</span><span>${localize('.dialog.expired-effects.content.message-continue')}<br>${formatTime(requestedNewWorldTime - newWorldTime)}?</span>`
+
+    promptUserToContinue(requestedNewWorldTime, newWorldTime, oldWorldTime, dialogTitle, dialogContent);
+  }
+
+  // Play the notification sound if a chat notification was made
+  const effectNotificationSoundSetting = game.settings.get(MODULE_ID, 'notifications-sound-for-expired-effects')
+  if (expireEffectMessageShown && effectNotificationSoundSetting.trim() !== '') {
+    foundry.audio.AudioHelper.play({src: effectNotificationSoundSetting});
+  }
+
+  args[0] = newWorldTime
+  wrapped(...args);
+}
+
+function promptUserToContinue(requestedNewWorldTime, newWorldTime, oldWorldTime, dialogTitle, dialogContent) {
+  new foundry.applications.api.DialogV2({
+    window: {
+      title: dialogTitle,
+      icon: 'fa-regular fa-alarm-exclamation'
+    },
+    content: dialogContent,
+    buttons: [{
+      icon: 'fas fa-check',
+      action: "yes",
+      label: localize('.dialog.expired-effects.buttons.yes'),
+      callback: async () => {
+        await game.time.advance(0) // stupid hack, without this the time doesn't advance properly?
+        await game.time.advance((requestedNewWorldTime - newWorldTime));
```

Comment:
If you can get this to work without the hack, I'd love to see what I did wrong?
I thought it may have been an async issue where advancing the time to exactly what the original time would be caused it but I tried all variations. (requestedNewWorldTime - newWorldTime) +1 or -1 would work just fine, so it has to be something with advancing 60 + 540 vs 600?

Which doesn't make sense to me as the user presses "yes" after the wrapper finishes? (unless it has something about calling the DialogV2 from a previous state?

I'm not a javascript "promise" expert at all :)